tools/buildsys/src/gomod.rs

```diff
@@ -0,0 +1,176 @@
+/*!
+Packages using the Go programming language may have upstream tar archives that
+include only the source code of the project, but not the source code of any
+dependencies. The Go programming language promotes the use of "modules" for
+dependencies and projects adopting modules will provide go.mod and go.sum
+files.
+
+This module extends the functionality of `packages.metadata.build-package.external-files`
+and provides the ability to retrieve and validate dependencies
+declared using Go modules given a tar archive containing a go.mod and go.sum.
+
+The location where dependencies are retrieved from are controlled by the
+standard environment variables employed by the Go tool: GOPROXY, GOSUMDB, and
+GOPRIVATE.
+
+ */
+
+pub(crate) mod error;
+use error::Result;
+
+use super::manifest;
+use duct::cmd;
+use snafu::{ensure, OptionExt, ResultExt};
+use std::env;
+use std::path::{Path, PathBuf};
+use std::process::Output;
+
+pub(crate) struct GoMod;
+
+impl GoMod {
+    pub(crate) fn vendor(
+        root_dir: &Path,
+        package_dir: &Path,
+        gomod: &manifest::ExternalFile,
+    ) -> Result<()> {
+        let url_file_name = extract_file_name(&gomod.url)?;
+        let path_arg = &gomod.path.as_ref().unwrap_or(&url_file_name);
+        ensure!(path_arg.components().count() == 1, error::InputFileSnafu);
+
+        let path = package_dir.join(path_arg);
+        ensure!(path.is_file(), error::InputFileBadSnafu { path });
+        let mod_dir = gomod.bundle_path.as_ref().context(error::ModDirSnafu)?;
+        let output_path_arg = gomod
+            .bundle_output_path
+            .as_ref()
+            .context(error::OutputDirSnafu)?;
+        let output_path = package_dir.join(output_path_arg);
+
+        // Return early if the output path exists and is a file, assuming it's already been built
+        if output_path.exists() && output_path.is_file() {
+            return Ok(());
+        }
```

Comment:
The discussion on single vs separate archives made me think of a potential failure mode here, which is that if the input external file changes in some way but keeps the same filename, we need to detect that and recreate the dependency bundle, rather than reusing the one we have.

If this optimization doesn't save very much time, it might be simpler to just always recreating the bundle file. I'm assuming that the `go mod vendor` step is pretty quick if all the modules are already downloaded into the cache.

Reply:
Does adding `rerun-if-changed` help here? Like rust's `Cargo.lock` file, the `go.mod` and `go.sum` files "lock" dependencies to a specific version. So there shouldn't be any change in what `go mod vendor` would download unless there are changes to those files.

At least I'm pretty sure. I've never mixed `go.mod` module management with vendoring in a project before, but this is the way I understand it to work.

Reply:
> it might be simpler to just always recreating the bundle file ... the go mod vendor step is pretty quick if all the modules are already downloaded into the cache.

I think recreating the dependency bundle is the way to go for now and should be almost instant given the modcahce, but I'll also do some experimenting to see if it slows down compared to my testing yesterday.

Reply:
> Does adding `rerun-if-changed` help here? Like rust's Cargo.lock file, the go.mod and go.sum files "lock" dependencies to a specific version.

It sort of helps, in the sense that any change to the file would trigger a rebuild - but if we're in `buildsys` at all then that means a (re)build has been triggered, and we don't know exactly why, just that `cargo` decided it was necessary. Since we're doing our own caching here, we'd need to detect whether the original input file changed - something else might've changed instead - and invalidate the bundled archive if it did.

That could be done by also saving a hash file of the original file somewhere, but then we have to manage that file too. Or we could prepend the hash of the original file to the bundled archive filename, so we could always check it, but then we'd want to strip off that prefix before passing into the package build environment since it wouldn't be predictable.

Not insurmountable problems but if the added benefit of the secondary cache is minor, it might not be worth the complexity.

Reply:
Ended up going with rebuilding the vendor directory and re-archiving. It didn't have an impact on performance since the local modcache should already be populated with the dependencies. It also seemed overly complicated to track the hashes of the dependencies and force a rebuild if any of those change.

But, like Sean said, I'd be very very surprised if those dependencies _did_ change out of band from the upstream releasing a new version of their software: the `go.mod` and `go.sum` act as locks on what dependencies are needed and shouldn't be different build to build.